lib/Dialect/Substrait/Transforms/EmitDeduplication.cpp

```diff
@@ -166,14 +220,61 @@ struct PushDuplicatesThroughCrossPattern : public OpRewritePattern<CrossOp> {
   }
 };
 
+/// Pushes duplicates in the mappings of `emit` ops producing the input through
+/// the `filter` op. This works by introducing a new `emit` op without the
+/// duplicates, creating a new `filter` op updated to work on the deduplicated
+/// element type, and finally a new `emit` op that maps back to the original
+/// order.
+struct PushDuplicatesThroughFilterPattern : public OpRewritePattern<FilterOp> {
+  using OpRewritePattern<FilterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(FilterOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!isa_and_present<EmitOp>(op.getInput().getDefiningOp()))
+      return rewriter.notifyMatchFailure(
+          op, "input operand is not produced by an 'emit' op");
+
+    // Create input ops for the new `filter` op. These may be the original
+    // inputs or `emit` ops that remove duplicates.
+    SmallVector<int64_t> reverseMapping;
+    auto [newInput, numDedupIndices, hasDuplicates] =
+        createDeduplicatingEmit(op.getInput(), reverseMapping, rewriter);
```

Comment:
Is this create a non-op unless hasDuplicates?

Reply:
Indeed, see docstring or note below: if `hasDuplicates == true`, then the call to `createDeduplicatingEmit` has no effect.